graphqlbackend: Deprecate URI again

keegancsmith opened the pull request:
Edit: This originally also added a fullName field, but we adjusted this PR to just remove URI until we are certain we want a fullName field.

URI is confusing and isn't used yet. Use the more descriptive fullName field. We intentionally do not update the DB type fields yet since that is a larger change. The more important change is how we adjust the user facing API.

Test plan: CI

Related https://github.com/sourcegraph/sourcegraph/pull/3995#pullrequestreview-245954335 and https://github.com/sourcegraph/sourcegraph/issues/4370

keegancsmith commented:
I originally exposed it since I thought it would be needed by browser or editor extensions. It may not, if so I'll adjust this PR to just not expose it at all.

I am unsure how you plan on solving https://github.com/sourcegraph/sourcegraph/issues/4368 but one possible way is to use the `fullName` rather than `name` when constructing URLs for the code host.

felixfbecker commented:
That is a possibility, although we currently use the Sourcegraph repo name. At some part we have to do the translation (if possible i.e. we have an instance), either in the beginning or when we want to jump to definition (since Sourcegraph browser URLs use the repo name).

How is the full name constructed/defined?

keegancsmith commented:
The full name is code host dependent. For most code hosts it is the "homepage" of the repo, without the scheme. It is the name with using the default repositoryPathPattern. eg github.com/owner/repo. The code for how we construct it is these search results https://sourcegraph.com/search?q=r:sourcegraph/sourcegraph%24+f:repo-updater/repos+-f:_test%5C.go%24+-f:store+%5CbURI:+

slimsag commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -1151,9 +1151,11 @@ type Repository implements Node & GenericSearchResultInterface {
+    # The full name for this repository (e.g., "github.com/user/repo").
     # See the documentation for the name field.
-    uri: String!
+    fullName: String!
```
Can you please clarify what the exact differences between `name` and `fullName` are? When should one be used and not the other? It is especially strange as a consumer because the example provided for full name exactly matches the "convention" described in the comment by `name`

slimsag commented:
I am surprised we want to introduce this instead of just getting rid of URI entirely and using the opaque `name` field everywhere.

felixfbecker commented:
@slimsag I had the same reaction, but am on the fence right now what to use for generating extension API URIs. Consider that in the browser extension, resolving the code host URL to a Sourcegraph repo `name` is one extra API request, which we cannot perform for private repos when we don't have a private Sourcegraph instance configured

felixfbecker commented:
> For most code hosts it is the "homepage" of the repo, without the scheme.

@keegancsmith what is it for other code hosts? Looked at the code but it wasn't obvious to me

keegancsmith commented:
> > For most code hosts it is the "homepage" of the repo, without the scheme.
> 
> @keegancsmith what is it for other code hosts? Looked at the code but it wasn't obvious to me

See examples from this integration test https://github.com/sourcegraph/sourcegraph/blob/cce50fcde5872e2c67f9769696104a8f69d9139c/cmd/repo-updater/repos/sources_test.go#L573-L641

Note: It is called URI here, but this is an implementation detail I can change. The URI here is returned as fullName in this PR.

keegancsmith commented:
> I am surprised we want to introduce this instead of just getting rid of URI entirely and using the opaque `name` field everywhere.

When on a code host an extension needs a way to construct the path for that code host. Previously it would rely on assumptions in name which break when repositoryPathPattern is configured. fullName allows the extension to make those assumptions without repositoryPathPattern concerns. However, I agree maybe there is a better API to expose here. Such as something related to our external links stuff.

felixfbecker commented:
@keegancsmith 	what's the use case you have in mind of an extension that needs to do that?

keegancsmith commented:
I don't know. Its just one approach one can take when resolving issues like https://github.com/sourcegraph/sourcegraph/issues/4368

This is just a proposal. I don't know what makes the most sense for the browser extension. How do you think extensions will resolve these sort of issues introduced by repositoryPathPattern?

keegancsmith commented:
@felixfbecker @slimsag I have removed the addition ot fullName, and updated this PR to just deprecate URI in graphql. I think we can both agree we should do that. We can leave discussion on fullName in #4370
